lib/services/local/desktops_test.go

```diff
@@ -0,0 +1,194 @@
+		Limit: 10,
+	})
+	require.NoError(t, err)
+	require.Empty(t, out.NextKey)
+	require.Empty(t, cmp.Diff([]types.WindowsDesktop{d1, d2, d3}, out.Desktops,
+		cmpopts.IgnoreFields(types.Metadata{}, "ID"),
+	))
+
+	t.Run("fetch first, middle, last", func(t *testing.T) {
+		t.Parallel()
+		// First
+		resp, err := service.ListWindowsDesktops(ctx, types.ListWindowsDesktopsRequest{
+			Limit: 1,
+		})
+		require.NoError(t, err)
+		require.Len(t, resp.Desktops, 1)
+		require.Equal(t, out.Desktops[0], resp.Desktops[0])
+		require.Equal(t, backend.GetPaginationKey(out.Desktops[1]), resp.NextKey)
+
+		// Middle
+		resp, err = service.ListWindowsDesktops(ctx, types.ListWindowsDesktopsRequest{
+			Limit:    1,
+			StartKey: resp.NextKey,
+		})
+		require.NoError(t, err)
+		require.Len(t, resp.Desktops, 1)
+		require.Equal(t, out.Desktops[1], resp.Desktops[0])
+		require.Equal(t, backend.GetPaginationKey(out.Desktops[2]), resp.NextKey)
+
+		// Last
+		resp, err = service.ListWindowsDesktops(ctx, types.ListWindowsDesktopsRequest{
+			Limit:    1,
+			StartKey: resp.NextKey,
+		})
+		require.NoError(t, err)
+		require.Len(t, resp.Desktops, 1)
+		require.Equal(t, out.Desktops[2], resp.Desktops[0])
+		require.Empty(t, resp.NextKey)
+	})
+
+	t.Run("filter", func(t *testing.T) {
+		t.Parallel()
+		tests := []struct {
+			name        string
+			filter      types.ListWindowsDesktopsRequest
+			wantErr     bool
+			expectedLen int
+		}{
+			{
+				name: "NOK non-matching host id and name",
+				filter: types.ListWindowsDesktopsRequest{
+					Limit: 10,
+					WindowsDesktopFilter: types.WindowsDesktopFilter{
+						HostID: "no-match",
+						Name:   "no-match",
+					},
+				},
+				wantErr: true,
+			},
+			{
+				name:    "NOK invalid limit",
+				filter:  types.ListWindowsDesktopsRequest{},
+				wantErr: true,
+			},
+			{
+				name: "matching host id",
+				filter: types.ListWindowsDesktopsRequest{
+					Limit:                1,
+					WindowsDesktopFilter: types.WindowsDesktopFilter{HostID: "test-host-id"},
+				},
+				expectedLen: 1,
+			},
+			{
+				name: "matching name",
+				filter: types.ListWindowsDesktopsRequest{
+					Limit:                1,
+					WindowsDesktopFilter: types.WindowsDesktopFilter{Name: "banana"},
+				},
+				expectedLen: 1,
+			},
+			{
+				name: "with search",
+				filter: types.ListWindowsDesktopsRequest{
+					Limit:          10,
+					SearchKeywords: []string{"env", "test"},
+				},
+				expectedLen: 2,
+			},
+			{
+				name: "with labels",
+				filter: types.ListWindowsDesktopsRequest{
+					Limit:  10,
+					Labels: testLabel,
+				},
+				expectedLen: 2,
+			},
+			{
+				name: "with predicate",
+				filter: types.ListWindowsDesktopsRequest{
+					Limit:               10,
+					PredicateExpression: `labels.env == "test"`,
+				},
+				expectedLen: 2,
+			},
+		}
+
+		for _, tc := range tests {
+			tc := tc
+			t.Run(tc.name, func(t *testing.T) {
+				t.Parallel()
```

Comment:
I'm not a big fan of subtests of subtests of subtests. Can we simplify this and just have more top-level tests?

Reply:
I separated the tests